daml-lf/tests/MultiKeys.daml

```diff
@@ -0,0 +1,232 @@
+-- Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module MultiKeys where
+
+import DA.Assert
+import DA.Exception (throw)
+import DA.Optional
+
+exception E
+  where
+    message "E"
+
+template Keyed
+  with
+    p : Party
+  where
+    signatory p
+    key p : Party
+    maintainer key
+
+-- All these tests operate under the assumption that there
+-- exists two global contracts with the same key `p` and one
+-- of those will always be returned by fetch/lookp-by-key at
+-- the beginning of the choice.
+-- All choices here are accepted in non-uck mode.
+-- Behavior in uck-mode is documented in comments.
+template KeyOperations
+  with
+    p : Party
+  where
+    signatory p
+
+    -- should be rejected in uck
+    nonconsuming choice CreateOverwritesLocal : ()
+      controller p
+      do cid1 <- create (Keyed p)
+         cid2 <- create (Keyed p)
+         (cid, _) <- fetchByKey @Keyed p
+         cid2 === cid
+
+    -- should be rejected in uck
+    nonconsuming choice CreateOverwritesUnknownGlobal : ()
+      controller p
+      do -- assume the global exists but do not fetch it
+         cid1 <- create (Keyed p)
+         (cid, _) <- fetchByKey @Keyed p
+         cid1 === cid
+
+    -- should be rejected in uck
+    nonconsuming choice CreateOverwritesKnownGlobal : ()
+      controller p
+      do _ <- fetchByKey @Keyed p
+         cid1 <- create (Keyed p)
+         (cid, _) <- fetchByKey @Keyed p
+         cid1 === cid
+
+    -- should be accepted in uck, cid not fetched by key
+    nonconsuming choice FetchDoesNotOverwriteGlobal : ()
+      with
+        cid : ContractId Keyed
+      controller p
+      do (cidByKey, _) <- fetchByKey @Keyed p
+         cid =/= cidByKey
+         c <- fetch cid
+         key c === p
+         (cidByKey', _) <- fetchByKey @Keyed p
+         cidByKey' === cidByKey
+
+    -- should be accepted in uck, cid not fetched by key
+    nonconsuming choice FetchDoesNotOverwriteLocal : ()
+      with
+        cid : ContractId Keyed
+      controller p
+      do local <- create (Keyed p)
+         c <- fetch cid
+         key c === p
+         (cid', _) <- fetchByKey @Keyed p
+         local === cid'
+
+    -- should be accepted in uck
+    nonconsuming choice LocalArchiveOverwritesUnknownGlobal : ()
+      controller p
+      do -- assume that a global contract with same key exists but don’t fetch it.
+         local <- create (Keyed p)
+         archive local
+         None <- lookupByKey @Keyed p
+         pure ()
+
+    -- should be rejected in uck
+    nonconsuming choice LocalArchiveOverwritesKnownGlobal : ()
+      controller p
+      do _ <- fetchByKey @Keyed p
+         local <- create (Keyed p)
+         archive local
+         None <- lookupByKey @Keyed p
```

Comment:
I'm on the fence w.r.t. the one. `lookupByKey` returning `None` is pretty meaningless in non-uck mode. Thinking about projections, it is probably the right choice, though:
The create-archive-lookup steps could be witnessed by some party `q` other than `p` and `q` doesn't witness the `fetchByKey`. So to `q`, this looks the same as `LocalArchiveOverwritesUnknownGlobal`.

Reply:
I'd say we should carefully revisit this once we've agreed on the semantics of non-uck key lookups. To me `lookupByKey` does not really make sense in non-uck mode, as its not clear at all what guarantee reading a `None` provides.

We might want to just forbid `lookupByKey` in non-uck mode.

Reply:
As discussed, we’re postponing multikey semantics so resolving this comment for this PR.